### DISCUSSION TITLE
Intercepting packet, and handling data transfer

### DISCUSSION THREAD

**Discussion by AlmightyLks** [chmorgan/sharppcap#458] [Q&A]
Doing what the docs shows
```cs
using var device = LibPcapLiveDeviceList.Instance[0];
device.Open();
device.OnPacketArrival += Device_OnPacketArrival;
device.StartCapture();

await Task.Delay(-1);

void Device_OnPacketArrival(object s, PacketCapture e)
{
    Console.WriteLine(e.Packet);
}
```

`Device_OnPacketArrival()` is called when packets arrive. This also counts for outbound packets.  
However, I wonder:  
You can manipulate and call `device.SendPacket()` manually; Is that what you're supposed to do? Or would calling it within the evenhandler cause it to send the same data twice? Especially, as _not_ calling it doesn't interfere with the network traffic...
The reason I am wondering about this, is that [NetStalker](https://github.com/hmz777/NetStalker/) does [exactly that](https://github.com/hmz777/NetStalker/blob/master/NetStalker/NetStalker/MainLogic/Blocker-Redirector.cs#L94). Following the same logic as NetStalker, that part of the code is responsible for rate-limitting. Meaning, when an outbound-packet comes in, check if its within the rate-limit, then send. Otherwise, don't call `device.SendPacket()`.

**Comment by kayoub5** [comment]
windivert driver is better suited for this than the npcap driver. 

That being said, there are registry entries that can be changed to make npcap cable of doing this, but they are only documented in npcap changelog.

**Reply by AlmightyLks** [reply]
Oh yeah, I was tackling WinDivert before, then I noticed NetStalker has accomplished this using sharppcap, which [also has a dependency on WinDivert](https://github.com/dotpcap/sharppcap/tree/master/SharpPcap/WinDivert) 😄 Was getting my hopes up, that this feature was already included here.
Thanks for the suggestion!
